# Самоучитель по Python для начинающих.

## Часть 17: Основы скрапинга и парсинга

In [7]:
# Импортируем нужные нам библиотеки.
import pandas as pd
from io import StringIO

import numpy as np
import random

In [74]:
# Импортируем нужные нам библиотеки.
from bs4 import BeautifulSoup
import requests

from urllib.request import urlopen

In [ ]:
# Импортируем нужные нам библиотеки.
from bs4 import BeautifulSoup
import requests

In [ ]:
# Импортируем нужные нам библиотеки.
from urllib.request import urlopen

In [50]:
# Использование объекта datetime :
from datetime import datetime

import time

In [ ]:
# Регулярные выражения
from urllib.request import urlopen
import re

In [ ]:
# Для работы с HTML кодом в Python используют модуль etree:
from urllib.request import urlopen
from lxml import etree

In [11]:
# Библиотека для имитации действий пользователя:

!pip install MechanicalSoup
# import mechanicalsoup

In [17]:
# Для работы с динамическим контентом в Python нужно установить:

# - Модуль Selenium.
# - Драйвер Selenium WebDriver для браузера.

from selenium import webdriver

In [36]:
# Библиотека для имитации действий пользователя:

# !pip install selenium
# import selenium
selenium.__version__

'3.14.0'

In [23]:
from selenium import webdriver

In [30]:
!pip install ChromeDriverManager
# ChromeDriverManager.__version__

In [31]:
ChromeDriverManager.__version__

NameError: name 'ChromeDriverManager' is not defined

In [28]:
!pip install webdriver_manager.chrome

ERROR: Could not find a version that satisfies the requirement webdriver_manager.chrome (from versions: none)
ERROR: No matching distribution found for webdriver_manager.chrome


In [32]:
from webdriver_manager.chrome import ChromeDriverManager

ModuleNotFoundError: No module named 'webdriver_manager'

In [44]:
# Бывает, что даже после установки оптимальной версии драйвера интерпретатор Python возвращает ошибку OSError:
# [WinError 216] Версия "%1" не совместима с версией Windows.
# В этом случае нужно воспользоваться модулем webdriver-manager,
# который самостоятельно установит подходящий драйвер для нужного браузера:

from selenium import webdriver
from selenium.webdriver.chrome.service import Service as ChromeService
from webdriver_manager.chrome import ChromeDriverManager
# driver = webdriver.Chrome(service=ChromeService(ChromeDriverManager().install()))

driver = webdriver.Chrome(ChromeDriverManager().install())

In [ ]:
# Use with Chrome
# selenium 3
# from selenium import webdriver
# from webdriver_manager.chrome import ChromeDriverManager

# driver = webdriver.Chrome(ChromeDriverManager().install())

In [ ]:
# selenium 4
# from selenium import webdriver
# from selenium.webdriver.chrome.service import Service as ChromeService
# from webdriver_manager.chrome import ChromeDriverManager

# driver = webdriver.Chrome(service=ChromeService(ChromeDriverManager().install()))

In [37]:
# Выполняем предыдущую ячейку ПОСТРОЧНО !!!
from selenium import webdriver

In [38]:
from selenium.webdriver.chrome.service import Service as ChromeService

In [40]:
# !pip install webdriver-manager
# from webdriver_manager.chrome import ChromeDriverManager

In [41]:
from webdriver_manager.chrome import ChromeDriverManager

In [42]:
driver = webdriver.Chrome(service=ChromeService(ChromeDriverManager().install()))

TypeError: __init__() got an unexpected keyword argument 'service'

Научимся извлекать данные из статического и динамического контента с помощью:
регулярных выражений, XPath, BeautifulSoup, MechanicalSoup и Selenium.

В конце статьи – код 10 скриптов для скрапинга данных и изображений с Wikipedia, Habr, LiveLib, IMDb и TIOBE.

Веб-скрапинг – это процесс автоматического сбора информации из онлайн-источников.
Для выбора нужных сведений из массива «сырых» данных, полученных в ходе скрапинга, нужна дальнейшая обработка – парсинг.
В процессе парсинга выполняются синтаксический анализ, разбор и очистка данных.
Результат парсинга – очищенные, упорядоченные, структурированные данные, представленные в формате, понятном конечному пользователю (или приложению).

Скрипты для скрапинга создают определенную нагрузку на сайт, с которого они собирают данные – могут, например, посылать чрезмерное количество GET запросов к серверу.
Это одна из причин, по которой скрапинг относится к спорным видам деятельности.
Чтобы не выходить за рамки сетевого этикета, необходимо всегда соблюдать главные правила сбора публичной информации:

- Если на сайте есть API, нужно запрашивать данные у него.
- Частота и количество GET запросов должны быть разумными.
- Следует передавать информацию о клиенте в User-Agent.
- Если на сайте есть личные данные пользователей, необходимо учитывать настройки приватности в robots.txt.

Необходимо отметить, что универсальных рецептов скрапинга и парсинга не существует.
Это связано со следующими причинами:

- Некоторые сервисы активно блокируют скраперов. Динамическая смена прокси не всегда помогает решить эту проблему.

- Контент многих современных сайтов генерируется динамически – результат обычного GET запроса из приложения к таким сайтам вернется практически пустым. Эта проблема решается с помощью Selenium WebDriver либо MechanicalSoup, которые имитируют действия браузера и пользователя.

Для извлечения данных со страниц с четкой, стандартной структурой эффективнее использовать язык запросов XPath.

И напротив, для получения нужной информации с нестандартных страниц с произвольным синтаксисом лучше использовать средства библиотеки BeautifulSoup.

Ниже мы подробно рассмотрим оба подхода.

### Примеры скрапинга и парсинга на Python

Экосистема Python располагает множеством инструментов для скрапинга и парсинга.
Начнем с самого простого примера – получения веб-страницы и извлечения из ее кода ссылки.

#### Скрапинг содержимого страницы

http://example.com/

In [1]:
# Воспользуемся модулем urllib.request стандартной библиотеки urllib
# для получения исходного кода одностраничного сайта example.com:
from urllib.request import urlopen

url = 'http://example.com'
page = urlopen(url)

print(page.read().decode('utf-8'))

<!doctype html>
<html>
<head>
    <title>Example Domain</title>

    <meta charset="utf-8" />
    <meta http-equiv="Content-type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type="text/css">
    body {
        background-color: #f0f0f2;
        margin: 0;
        padding: 0;
        font-family: -apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
        
    }
    div {
        width: 600px;
        margin: 5em auto;
        padding: 2em;
        background-color: #fdfdff;
        border-radius: 0.5em;
        box-shadow: 2px 3px 7px 2px rgba(0,0,0,0.02);
    }
    a:link, a:visited {
        color: #38488f;
        text-decoration: none;
    }
    @media (max-width: 700px) {
        div {
            margin: 0 auto;
            width: auto;
        }
    }
    </style>    
</head>

<body>
<div>
    <h1>Example Domain</h1>
    <p>This domai

In [2]:
# Точно такой же результат можно получить с помощью requests:
from bs4 import BeautifulSoup
import requests

url = 'http://example.com'
res = requests.get(url)
soup = BeautifulSoup(res.text,'html.parser')

print(soup)

<!DOCTYPE html>

<html>
<head>
<title>Example Domain</title>
<meta charset="utf-8"/>
<meta content="text/html; charset=utf-8" http-equiv="Content-type"/>
<meta content="width=device-width, initial-scale=1" name="viewport"/>
<style type="text/css">
    body {
        background-color: #f0f0f2;
        margin: 0;
        padding: 0;
        font-family: -apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
        
    }
    div {
        width: 600px;
        margin: 5em auto;
        padding: 2em;
        background-color: #fdfdff;
        border-radius: 0.5em;
        box-shadow: 2px 3px 7px 2px rgba(0,0,0,0.02);
    }
    a:link, a:visited {
        color: #38488f;
        text-decoration: none;
    }
    @media (max-width: 700px) {
        div {
            margin: 0 auto;
            width: auto;
        }
    }
    </style>
</head>
<body>
<div>
<h1>Example Domain</h1>
<p>This domain is for use in illustrative example

Этот результат – те самые сырые данные, которые нужно обработать (подвергнуть парсингу), чтобы извлечь из них нужную информацию, например, адрес указанной на странице ссылки:

#### Парсинг полученных данных

Извлечь адрес ссылки можно 4 разными способами – с помощью:

- Методов строк.
- Регулярного выражения.
- Запроса XPath.
- Обработки BeautifulSoup.

Рассмотрим все эти способы по порядку.

In [ ]:
# 

#### Методы строк

Это самый трудоемкий способ – для извлечения каждого элемента нужно определить 2 индекса – начало и конец вхождения.
При этом к индексу вхождения надо добавить длину стартового фрагмента:

In [4]:
from urllib.request import urlopen

# 
url = 'http://example.com/'
page = urlopen(url)
html_code = page.read().decode('utf-8')

# 
start = html_code.find('href="') + 6
end = html_code.find('">More')
link = html_code[start:end]

print(start)
print(end)

print(link)

1167
1203
https://www.iana.org/domains/example


In [ ]:
# 

#### Регулярное выражение

В предыдущей главе мы подробно рассматривали способы извлечения конкретных подстрок из текста.
https://proglib.io/p/samouchitel-po-python-dlya-nachinayushchih-chast-16-regulyarnye-vyrazheniya-2023-02-20

Точно так же регулярные выражения можно использовать для поиска данных в исходном коде страниц:

In [5]:
from urllib.request import urlopen
import re

# 
url = 'http://example.com/'
page = urlopen(url)
html_code = page.read().decode('utf-8')

# 
link = r'(https?://\S+)(?=")'

print(re.findall(link, html_code))

['https://www.iana.org/domains/example']


In [ ]:
# 

#### Запрос XPath

Язык запросов XPath (XML Path Language) позволяет извлекать данные из определенных узлов XML-документа.
Для работы с HTML кодом в Python используют модуль etree:

In [6]:
from urllib.request import urlopen
from lxml import etree

# 
url = 'http://example.com/'
page = urlopen(url)
html_code = page.read().decode('utf-8')
tree = etree.HTML(html_code)

print(tree.xpath("/html/body/div/p[2]/a/@href")[0])

https://www.iana.org/domains/example


Чтобы узнать путь к нужному элементу страницы, в браузерах Chrome и FireFox надо кликнуть правой кнопкой по элементу и выбрать «Просмотреть код», после чего откроется консоль.
В консоли по интересующему элементу нужно еще раз кликнуть правой кнопкой, выбрать «Копировать», а затем – копировать путь XPath:

В приведенном выше примере для извлечения ссылки к пути /html/body/div/p[2]/a/
мы добавили указание для получения значения ссылки @href, и индекс [0], поскольку результат возвращается в виде списка.

###### Если @href заменить на text(), программа вернет текст ссылки, а не сам URL:

In [7]:
print(tree.xpath("/html/body/div/p[2]/a/text()")[0])

More information...


### Парсер BeautifulSoup

Регулярные выражения и XPath предоставляют огромные возможности для извлечения нужной информации из кода страниц, но у них есть свои недостатки:
составлять Regex-шаблоны сложно, а запросы XPath хорошо работают только на страницах с безупречной, стандартной структурой.
К примеру, страницы Википедии не отличаются идеальной структурой, и использование XPath для извлечения нужной информации из определенных элементов статей, таких как таблицы infobox, часто оказывается неэффективным.

##### В этом случае оптимальным вариантом становится BeautifulSoup, специально разработанный для парсинга HTML-кода.

In [ ]:
# Библиотека BeautifulSoup не входит в стандартный набор Python, ее нужно установить самостоятельно:
        
# pip install beautifulsoup4

В приведенном ниже примере мы будем извлекать из исходного кода страницы уникальные ссылки, за исключением внутренних:

https://webscraper.io/test-sites/e-commerce/allinone/phones

In [8]:
from bs4 import BeautifulSoup
from urllib.request import urlopen

# 
url = 'https://webscraper.io/test-sites/e-commerce/allinone/phones'
page = urlopen(url)
html = page.read().decode('utf-8')
soup = BeautifulSoup(html, 'html.parser')

# 
links = set()
for link in soup.find_all('a'):
    l = link.get('href')
    if l != None and l.startswith('https'):
        links.add(l)
for link in links:
    print(link)

https://twitter.com/webscraperio
https://forum.webscraper.io/
https://cloud.webscraper.io/
https://www.facebook.com/webscraperio/
https://chromewebstore.google.com/detail/web-scraper-free-web-scra/jnhgnonknehpejjnehehllkliplmbmhn?hl=en
https://youtube.com/@WebScraper/videos
https://webscraper.io/downloads/Web_Scraper_Media_Kit.zip
https://lv.linkedin.com/company/web-scraper
https://status.webscraper.io/


In [9]:
# При использовании XPath точно такой же результат даст следующий скрипт:
from urllib.request import urlopen
from lxml import etree

# 
url = 'https://webscraper.io/test-sites/e-commerce/allinone/phones'
page = urlopen(url)
html_code = page.read().decode('utf-8')
tree = etree.HTML(html_code)
sp = tree.xpath("//li/a/@href")

# 
links = set()
for link in sp:
    if link.startswith('http'):
        links.add(link)

for link in links:
    print(link)

https://twitter.com/webscraperio
https://forum.webscraper.io/
https://cloud.webscraper.io/
https://www.facebook.com/webscraperio/
https://chromewebstore.google.com/detail/web-scraper-free-web-scra/jnhgnonknehpejjnehehllkliplmbmhn?hl=en
https://youtube.com/@WebScraper/videos
https://webscraper.io/downloads/Web_Scraper_Media_Kit.zip
https://lv.linkedin.com/company/web-scraper
https://status.webscraper.io/


### Имитация действий пользователя в браузере

При скрапинге сайтов очень часто требуется авторизация, нажатие кнопок «Читать дальше», переход по ссылкам, отправка форм, прокручивание ленты и так далее.
Отсюда возникает необходимость имитации действий пользователя.
Как правило, для этих целей используют Selenium, однако есть и более легкое решение – библиотека MechanicalSoup:

###### pip install MechanicalSoup

По сути, MechanicalSoup исполняет роль браузера без графического интерфейса.
Помимо имитации нужного взаимодействия с элементами страниц,
MechanicalSoup также парсит HTML-код, используя для этого все функции BeautifulSoup.

Воспользуемся тестовым сайтом http://httpbin.org/, на котором есть возможность отправки формы заказа пиццы:

In [12]:
import mechanicalsoup

# 
browser = mechanicalsoup.StatefulBrowser()
browser.open("http://httpbin.org/")
browser.follow_link("forms")
browser.select_form('form[action="/post"]')

print(browser.form.print_summary())

# В приведенном выше примере браузер MechanicalSoup перешел по внутренней ссылке http://httpbin.org/forms/post
# и вернул описание полей ввода:

<input name="custname"/>
<input name="custtel" type="tel"/>
<input name="custemail" type="email"/>
<input name="size" type="radio" value="small"/>
<input name="size" type="radio" value="medium"/>
<input name="size" type="radio" value="large"/>
<input name="topping" type="checkbox" value="bacon"/>
<input name="topping" type="checkbox" value="cheese"/>
<input name="topping" type="checkbox" value="onion"/>
<input name="topping" type="checkbox" value="mushroom"/>
<input max="21:00" min="11:00" name="delivery" step="900" type="time"/>
<textarea name="comments"></textarea>
<button>Submit order</button>
None


In [13]:
# Перейдем к имитации заполнения формы:
        
browser["custname"] = "Best Customer"
browser["custtel"] = "+7 916 123 45 67"
browser["custemail"] = "trex@example.com"
browser["size"] = "large"
browser["topping"] = ("cheese", "mushroom")
browser["comments"] = "Add more cheese, plz. More than the last time!"

In [14]:
# Теперь форму можно отправить:
        
response = browser.submit_selected()

In [15]:
# Результат можно вывести с помощью print(response.text):
print(response.text)

{
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {
    "comments": "Add more cheese, plz. More than the last time!", 
    "custemail": "trex@example.com", 
    "custname": "Best Customer", 
    "custtel": "+7 916 123 45 67", 
    "delivery": "", 
    "size": "large", 
    "topping": [
      "cheese", 
      "mushroom"
    ]
  }, 
  "headers": {
    "Accept": "*/*", 
    "Accept-Encoding": "gzip, deflate", 
    "Content-Length": "191", 
    "Content-Type": "application/x-www-form-urlencoded", 
    "Host": "httpbin.org", 
    "Referer": "http://httpbin.org/forms/post", 
    "User-Agent": "python-requests/2.24.0 (MechanicalSoup/1.3.0)", 
    "X-Amzn-Trace-Id": "Root=1-67155811-78bd5d122a1637122c7af07f"
  }, 
  "json": null, 
  "origin": "94.25.225.163", 
  "url": "http://httpbin.org/post"
}



### Скрапинг и парсинг динамического контента

Все примеры, которые мы рассмотрели выше, отлично работают на статических страницах.
Однако на множестве платформ используется динамический подход к генерации и загрузке контента –
к примеру, для просмотра всех доступных товаров в онлайн-магазине страницу нужно не только открыть, но и прокрутить до футера. Для работы с динамическим контентом в Python нужно установить:

- Модуль Selenium.
- Драйвер Selenium WebDriver для браузера.

In [48]:
# $$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$ ИСПОЛЬЗОВАТЬ  ДАННЫЙ  КОД  ДЛЯ  ПАРСИНГА  САЙТОВ $$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$

# Use with Chrome
# selenium 3

# from selenium import webdriver
# from webdriver_manager.chrome import ChromeDriverManager

driver = webdriver.Chrome(ChromeDriverManager().install())

# $$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$ ИСПОЛЬЗОВАТЬ  ДАННЫЙ  КОД  ДЛЯ  ПАРСИНГА  САЙТОВ $$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$

In [49]:
# Если установка прошла успешно, выполнение этого кода приведет к автоматическому открытию страницы:
# from selenium import webdriver


# driver = webdriver.Chrome() # или webdriver.Firefox()
driver = webdriver.Chrome(ChromeDriverManager().install())
driver.get('https://google.com')

#### Имитация прокрутки страницы и парсинг данных

В качестве примера скрапинга и парсинга динамического сайта мы воспользуемся разделом тестового онлайн-магазина.
Здесь расположены карточки с информацией о планшетах.
Карточки загружаются ряд за рядом при прокрутке страницы:

https://webscraper.io/test-sites/e-commerce/scroll/computers/tablets

Пока страница не прокручена, полный HTML-код с информацией о планшетах получить невозможно.
Для имитации прокрутки мы воспользуемся скриптом 'window.scrollTo(0, document.body.scrollHeight);'.
Цены планшетов находятся в тегах "H4" класса pull-right price, а названия моделей – в тексте ссылок "A" класса title.
Готовый код выглядит так:

In [60]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service as ChromeService
from webdriver_manager.chrome import ChromeDriverManager
import time
from bs4 import BeautifulSoup

# Оригинальная функция к данной задаче 
# driver = webdriver.Chrome(service=ChromeService(ChromeDriverManager(cache_valid_range=10).install()))

# Функция полученная несколько ячеек выше с добавлением "cache_valid_range=10":
# driver = webdriver.Chrome(ChromeDriverManager().install())
# driver = webdriver.Chrome(ChromeDriverManager(cache_valid_range=10).install())


url = 'https://webscraper.io/test-sites/e-commerce/scroll/computers/tablets'
driver.get(url)
driver.execute_script('window.scrollTo(0, document.body.scrollHeight);')

time.sleep(5) 
html = driver.page_source
soup = BeautifulSoup(html, 'html.parser')
prices = soup.find_all('h4', class_='pull-right price')
models = soup.find_all('a', class_='title')

for model, price in zip(models, prices):
    m = model.get_text()
    p = price.get_text()
    print(f'Планшет {m}, цена - {p}')

WebDriverException: Message: 'chromedriver.exe' executable needs to be in PATH. Please see https://sites.google.com/a/chromium.org/chromedriver/home


### Практика

#### Задание 1

Напишите программу для получения названий последних статей из блога издательства O’Reilly:

https://www.oreilly.com/radar/

In [61]:
# from bs4 import BeautifulSoup
# import requests

url = 'https://www.oreilly.com/radar/'
headers = {"User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.96 Safari/537.36"}

res = requests.get(url, headers=headers)
soup = BeautifulSoup(res.text,'html.parser')
sp = soup.find_all('h2', class_='post-title')
for post in sp:
    print(post.get_text())

Generative AI in the Real World: The Future of Programming with Matt Welsh
The State of Security in 2024
Beyond Imitation
Henry Ford Does AI


In [ ]:
# 

#### Задание 2

Напишите программу для определения 10 слов, которые чаще всего встречаются в тексте сказки «Колобок» (без учета регистра). Предлоги учитывать не нужно.

https://azku.ru/russkie-narodnie-skazki/kolobok.html

In [62]:
# from bs4 import BeautifulSoup
# import requests
# import re

url = 'https://azku.ru/russkie-narodnie-skazki/kolobok.html'
res = requests.get(url)
soup = BeautifulSoup(res.text,'html.parser')
res = soup.find_all('div', class_='entry-content')
text = res[0].get_text().split('Мне нравится')[0]
sp = re.sub(r'[-,.?!"—:]', ' ', text).lower().split()

result_dict = {word: sp.count(word) for word in sp if len(word) > 3}
max_values = sorted(result_dict.items(), key=lambda x:x[1], reverse=True)[:10]

for word, number in max_values:
    print(f'{word} - {number}')

колобок - 32
ушел - 14
тебя - 7
коробу - 6
сусеку - 6
сметане - 5
масле - 5
покатился - 4
катится - 4
навстречу - 4


In [ ]:
# 

#### Задание 3

Напишите программу, которая на основе данных таблицы создает список цитат из фильмов, выпущенных после 1995 года.

https://ru.wikipedia.org/wiki/100_%D0%B8%D0%B7%D0%B2%D0%B5%D1%81%D1%82%D0%BD%D1%8B%D1%85_%D1%86%D0%B8%D1%82%D0%B0%D1%82_%D0%B8%D0%B7_%D0%B0%D0%BC%D0%B5%D1%80%D0%B8%D0%BA%D0%B0%D0%BD%D1%81%D0%BA%D0%B8%D1%85_%D1%84%D0%B8%D0%BB%D1%8C%D0%BC%D0%BE%D0%B2_%D0%B7%D0%B0_100_%D0%BB%D0%B5%D1%82_%D0%BF%D0%BE_%D0%B2%D0%B5%D1%80%D1%81%D0%B8%D0%B8_AFI

In [63]:
# from bs4 import BeautifulSoup
# import requests
# import re

url = 'https://ru.wikipedia.org/wiki/100_%D0%B8%D0%B7%D0%B2%D0%B5%D1%81%D1%82%D0%BD%D1%8B%D1%85_%D1%86%D0%B8%D1%82%D0%B0%D1%82_%D0%B8%D0%B7_%D0%B0%D0%BC%D0%B5%D1%80%D0%B8%D0%BA%D0%B0%D0%BD%D1%81%D0%BA%D0%B8%D1%85_%D1%84%D0%B8%D0%BB%D1%8C%D0%BC%D0%BE%D0%B2_%D0%B7%D0%B0_100_%D0%BB%D0%B5%D1%82_%D0%BF%D0%BE_%D0%B2%D0%B5%D1%80%D1%81%D0%B8%D0%B8_AFI'
res = requests.get(url).text
soup = BeautifulSoup(res,'html.parser')
table = soup.find('table', class_='wikitable')
header = table.find_all('th')
quotes = table.find_all('tr')
quotes_after_1995 = []

for i in range(1, len(quotes)):
    q = quotes[i].get_text()
    if int(q[-5:]) > 1995:
        quotes_after_1995.append(' '.join(q.split('\n\n')[1:]).replace('\n', ''))
        
for quote in quotes_after_1995:
    print(quote)

Show me the money! Покажи мне деньги! Род Тидвелл Кьюба Гудинг мл. Джерри Магуайер 1996
I see dead people. Я вижу мёртвых людей. Коул Сиэр Хэйли Джоэл Осмент Шестое чувство 1999
You had me at 'hello'. Я была твоя уже на «здрасьте». Дороти Бойд Рене Зеллвегер Джерри Магуайер 1996
My precious. Моя прелесть. Голлум Энди Серкис Властелин колец: Две крепости 2002
I’m the king of the world! Я король мира! Джек Доусон Леонардо Ди Каприо Титаник 1997


In [ ]:
# 

#### Задание 4

Напишите программу, которая извлекает данные о моделях, конфигурации и стоимости 117-ти ноутбуков, и записывает полученную информацию в csv файл.

https://webscraper.io/test-sites/e-commerce/allinone/computers/laptops

### Константы для проекта

In [24]:
# Создаём переменные-КОНСТАНТЫ
# 1. Адрес основного хоста
HOST = ''
# https://chelyabinsk.bani.ru.com/poleznye-statiy/pravila-ekspluatacii-bani-bochki-kak-prodlit-srok-sluzhby/

# 2. Адрес страницы для парсинга
URL = 'https://webscraper.io/test-sites/e-commerce/allinone/computers/laptops'

# 3. Создаём словарь заголовков
# Это нужно для некоторых сайтов (но не для всех - для этого необязательно)
# чтобы показать сайту, что ВЫ НЕ БОТ, а обычный пользователь
HEADERS = {
    'accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
    'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36'
}

In [27]:
# Отправляем запрос на страницу, передаём в качестве GET-параметра URL-адрес
# Сохраним нашу страницу в файл И ЗАКОММЕНТИРУЕМ ЭТУ КОМАНДУ за ненадобностью

# res = requests.get(url=URL, headers=HEADERS)
# src = res.text
# with open('../data/interim/html/index_laptop.html', 'w', encoding='utf-8') as file:
#     file.write(src)

# Теперь откроем наш файл, прочитаем и сохраним код страницы в переменную:
with open('../data/interim/html/index_laptop.html', encoding='utf-8') as file:
    src = file.read()
# print(src)

# f'data/{count}_{category_name}.csv'

In [29]:
# Чтобы пользоваться методами библиотеки BEAUTIFULSOUP нужно открыть нашу переменную в этой библиотеке
# (Преобразовать код в дерево объектов Python)
soup = BeautifulSoup(src, "lxml")

# LXML - это название название парсера (считается самым быстрым)
# В стандартной установке его нет (надо устанавливать отдельно)
# !pip install lxml

In [30]:
# url = 'https://webscraper.io/test-sites/e-commerce/allinone/computers/laptops'
# res = requests.get(src)
# soup = BeautifulSoup(res.text,'html.parser')
models = soup.find_all('a', class_='title')
description = soup.find_all('p', class_='description')
prices = soup.find_all('h4', class_='pull-right price')

In [39]:
models = list(soup.find_all('a', class_='title'))
description = list(soup.find_all('p', class_='description'))
prices = list(soup.find_all('h4', class_='pull-right price'))

In [40]:
# list(models)
type(models)

list

In [43]:
print(f'Модель;Описание;Цена\n')
for m, d, p in zip(models, description, prices):
    print(f"{m};{d};{p}\n")

Модель;Описание;Цена



In [34]:
print(f'Модель;Описание;Цена\n')
for m in models:
    print(f"{m['title']}")

Модель;Описание;Цена

Asus VivoBook X441NA-GA190
Prestigio SmartBook 133S Dark Grey
Prestigio SmartBook 133S Gold
Aspire E1-510
Lenovo V110-15IAP
Lenovo V110-15IAP
Hewlett Packard 250 G6 Dark Ash Silver
Acer Aspire 3 A315-31 Black
Acer Aspire A315-31-C33J
Acer Aspire ES1-572 Black
Acer Aspire 3 A315-31 Black
Acer Aspire 3 A315-21
Asus VivoBook Max
Asus VivoBook E502NA-GO022T Dark Blue
Lenovo ThinkPad E31-80
Acer Aspire 3 A315-31 Black
Lenovo V110-15ISK
Acer Aspire ES1-732 Black
Asus VivoBook 15 X540NA-GQ026T
Packard 255 G2
Asus EeeBook R416NA-FA014T
Acer Aspire 3 A315-51
Acer Aspire ES1-572 Black
Acer Extensa 15 (2540) Black
Acer Aspire ES1-572 Black
Lenovo V110-15ISK
Acer Aspire A315-51-33TG
Lenovo V110-15IKB
Asus VivoBook 15 X540UA-DM260 Chocolate Black
Acer Aspire ES1-572 Black
Lenovo V510 Black
Acer Aspire ES1-572 Black
Lenovo V510 Black
Acer Swift 1 SF113-31 Silver
Dell Vostro 15
Acer Aspire 3 A315-51 Black
Dell Vostro 15 (3568) Red
Lenovo V510 Black
HP 250 G3
Acer Spin 5
HP 350 G

In [38]:
print(f'Модель;Описание;Цена\n')
for m, d, p in zip(list(models), list(description), list(prices)):
    print(f"{m['title']};{d.get_text()};{p.get_text()}\n")

Модель;Описание;Цена



In [5]:
# from bs4 import BeautifulSoup
# import requests

url = 'https://webscraper.io/test-sites/e-commerce/allinone/computers/laptops'
res = requests.get(url)
soup = BeautifulSoup(res.text,'html.parser')
models = soup.find_all('a', class_='title')
description = soup.find_all('p', class_='description')
prices = soup.find_all('h4', class_='pull-right price')

with open('../data/interim/csv/laptops.csv', 'w', encoding='utf-8') as file:
    # file.write(f'Модель;Описание;Цена\n')
    print(f'Модель;Описание;Цена\n')
    for m, d, p in zip(models, description, prices):
        # file.write(f"{m['title']};{d.get_text()};{p.get_text()}\n")
        print(f"{m};{d};{p}\n")
        file.write(f"{m};{d};{p}\n")

In [22]:
description[0]

<p class="description card-text">Asus VivoBook X441NA-GA190 Chocolate Black, 14", Celeron N3450, 4GB, 128GB SSD, Endless OS, ENG kbd</p>

In [8]:
# Проверка выполненного задания - загружаем данные из файла CSV
laptops = pd.read_csv('../data/interim/csv/laptops.csv')
laptops.head()

Empty DataFrame
Columns: [Модель;Описание;Цена]
Index: []

#### Задание 5

Напишите программу для скачивания полноразмерных обложек из профилей книг на LiveLib.
Обложки открываются после двойного клика по миниатюре:

https://www.livelib.ru/book/1002978643-ohotnik-za-tenyu-donato-karrizi

In [71]:
# from bs4 import BeautifulSoup
# import requests
# import re

url = 'https://www.livelib.ru/book/1002978643-ohotnik-za-tenyu-donato-karrizi'
headers = {"User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.96 Safari/537.36"}

res = requests.get(url, headers=headers)
soup = BeautifulSoup(res.text,'html.parser')
sp = soup.find('div', class_='bc-menu__image-wrapper')
img_url = re.findall(r'(?:https\:)?//.*\.(?:jpeg)', str(sp))[0]
response = requests.get(img_url, headers=headers)

if response.status_code == 200:
    file_name = url.split('-', 1)[1]
    with open(file_name + '.jpeg', 'wb') as file:
        file.write(response.content)

IndexError: list index out of range

In [ ]:
# 

#### Задание 6

Напишите программу, которая составляет рейтинг топ-100 лучших триллеров на основе этого списка.

https://www.livelib.ru/genre/%D0%A2%D1%80%D0%B8%D0%BB%D0%BB%D0%B5%D1%80%D1%8B/top

In [75]:
# from bs4 import BeautifulSoup
# import requests

url = 'https://www.livelib.ru/genre/%D0%A2%D1%80%D0%B8%D0%BB%D0%BB%D0%B5%D1%80%D1%8B/top'
headers = {"User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.96 Safari/537.36"}

res = requests.get(url, headers=headers)
soup = BeautifulSoup(res.content,'html.parser')
titles = soup.find_all('a', class_='brow-book-name with-cycle')
authors = soup.find_all('a', class_='brow-book-author')
rating = soup.find_all('span', class_='rating-value stars-color-orange')

i = 1
for t, a, r in zip(titles, authors, rating):
    print(f'{i}. "{t.get_text()}", {a.get_text()} - {r.get_text()}')
    i += 1

In [ ]:
# 

#### Задание 7

Напишите программу, которая составляет топ-20 языков программирования на основе рейтинга популярности TIOBE.

https://www.tiobe.com/tiobe-index/

In [76]:
import requests
from lxml import html

url = 'https://www.tiobe.com/tiobe-index/'
headers = {"User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.96 Safari/537.36"}

page = requests.get(url, headers=headers)
tree = html.fromstring(page.content)
languages, rating = [], []

for i in range(1, 21):
    languages.append(tree.xpath(f'//*[@id="top20"]/tbody/tr[{i}]/td[5]/text()')[0])
    rating.append(tree.xpath(f'//*[@id="top20"]/tbody/tr[{i}]/td[6]/text()')[0])
i = 1

for l, r in zip(languages, rating):
    print(f'{i}. {l}: {r}')
    i += 1

1. Python: 21.90%
2. C++: 11.60%
3. Java: 10.51%
4. C: 8.38%
5. C#: 5.62%
6. JavaScript: 3.54%
7. Visual Basic: 2.35%
8. Go: 2.02%
9. Fortran: 1.80%
10. Delphi/Object Pascal: 1.68%
11. SQL: 1.64%
12. MATLAB: 1.48%
13. Rust: 1.45%
14. Scratch: 1.41%
15. PHP: 1.21%
16. Assembly language: 1.13%
17. R: 1.09%
18. Ruby: 0.99%
19. COBOL: 0.99%
20. Swift: 0.98%


In [ ]:
# 

#### Задание 8

Напишите программу для получения рейтинга 250 лучших фильмов по версии IMDb. Названия должны быть на русском языке.

https://www.imdb.com/chart/top/

In [77]:
import requests
from lxml import html

url = 'https://www.imdb.com/chart/top/'
headers = {"User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.96 Safari/537.36",
           "Accept-Language": "ru-RU"}

page = requests.get(url, headers=headers)
tree = html.fromstring(page.content)
movies, year, rating = [], [], []

for i in range(1, 251):
    movies.append(tree.xpath(f'//*[@id="main"]/div/span/div/div/div[3]/table/tbody/tr[{i}]/td[2]/a/text()')[0])
    year.append(tree.xpath(f'//*[@id="main"]/div/span/div/div/div[3]/table/tbody/tr[{i}]/td[2]/span/text()')[0])
    rating.append(tree.xpath(f'//*[@id="main"]/div/span/div/div/div[3]/table/tbody/tr[{i}]/td[3]/strong/text()')[0])
i = 1

for m, y, r in zip(movies, year, rating):
    print(f'{i}. {m}, {y}, {r}')
    i += 1

IndexError: list index out of range

In [ ]:
# 

#### Задание 9

Напишите программу, которая сохраняет в текстовый файл данные о фэнтези фильмах с 10 первых страниц соответствующего раздела IMDb.
Если у фильма/сериала еще нет рейтинга, следует указать N/A.

https://www.imdb.com/search/title/?genres=fantasy

Ожидаемый результат в файле fantasy.txt – 500 записей:

In [78]:
import requests
import mechanicalsoup
from lxml import html
import time

url = 'https://www.imdb.com/search/title/?genres=fantasy'
headers = {"User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.96 Safari/537.36",
           "Accept-Language": "ru-RU"}

browser = mechanicalsoup.StatefulBrowser()
j = 51

for _ in range(10):
    browser.open(url)
    page = requests.get(url, headers=headers)
    tree = html.fromstring(page.content)
    titles, year, rating = [], [], []
    for i in range(1, 51):
        if tree.xpath(f'//*[@id="main"]/div/div[3]/div/div[{i}]/div[3]/p[1]/b/text()') != []:
            titles.append(tree.xpath(f'//*[@id="main"]/div/div[3]/div/div[{i}]/div[3]/h3/a/text()')[0])
            year.append(tree.xpath(f'//*[@id="main"]/div/div[3]/div/div[{i}]/div[3]/h3/span[2]/text()')[0])
            rating.append('N/A')
        else:
            titles.append(tree.xpath(f'//*[@id="main"]/div/div[3]/div/div[{i}]/div[3]/h3/a/text()')[0])
            year.append(tree.xpath(f'//*[@id="main"]/div/div[3]/div/div[{i}]/div[3]/h3/span[2]/text()')[0])
            rating.append(tree.xpath(f'//*[@id="main"]/div/div[3]/div/div[{i}]/div[3]/div/div[1]/strong/text()')[0])
    with open('fantasy.txt', 'a', encoding='utf-8') as file:
        for t, y, r in zip(titles, year, rating):
            file.write(f'{t}, {y}, {r}\n')

    time.sleep(2)    
    lnk = browser.follow_link('start=' + str(j))
    url = browser.url
    j += 50

IndexError: list index out of range

In [ ]:
# 

#### Задание 10

Напишите программу для получения главных новостей (на русском) с портала Habr.
Каждый заголовок должен сопровождаться ссылкой на полный текст новости.

https://habr.com/ru/news/

In [1]:
from bs4 import BeautifulSoup
import requests

url = 'https://habr.com/ru/news/'
headers = {"User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.96 Safari/537.36",
           "Accept-Language": "ru-RU"}
res = requests.get(url, headers=headers)
soup = BeautifulSoup(res.content,'html.parser')
articles = soup.find_all('a', class_='tm-article-snippet__title-link')

for a in articles:
    print(f'{a.get_text()}\nhttps://habr.com{a.get("href")}')

In [ ]:
# 

#### Задание 11

In [ ]:
# 

In [ ]:
# 

In [ ]:
# 